pkg/infra/httpclient/httpclientprovider/set_headers_middleware.go

```diff
@@ -21,8 +24,14 @@ func SetHeadersMiddleware(headers http.Header) httpclient.Middleware {
 
 		return httpclient.RoundTripperFunc(func(req *http.Request) (*http.Response, error) {
 			for k, v := range headers {
-				canonicalKey := textproto.CanonicalMIMEHeaderKey(k)
-				req.Header[canonicalKey] = v
+				key := textproto.CanonicalMIMEHeaderKey(k)
+				if key == textproto.CanonicalMIMEHeaderKey(idTokenHeaderName) {
```

Comment:
Maybe add logic checking if any header is already set and in that case skip adding/setting it? That would be similar to how other middlewares work, e.g. https://github.com/grafana/grafana/blob/9a41277ecdbf327fadb6acfbf0e3415098d6503c/pkg/infra/httpclient/httpclientprovider/user_agent_middleware.go#L21-L24

Reply:
I believe it's a good idea!

The root cause of #59940 is the Prometheus datasource setting the headers without canonicalizing it first (I have opened #60342 which fixes it), however there may be more core plugins that do a similar thing, so if we do it at middleware level it will fix the issue for all of them.

It's a HTTP thing but the issue seem to be _mainly_ for `X-Id-Token` vs `X-ID-Token` (so OAuth), most likely because ID makes sense in English but it becomes Id in MIME canonical form. Confusing :exploding_head:  !

I tried adding the headers existence check like you have suggested to remove the dependency on `X-ID-Token`, but unfortunately it doesn't solve the issue because the wrong `X-ID-Token` is set before calling the middleware, and if we canonicalize it we get `X-Id-Token`, which doesn't exist, so the check fails.

At this point I think it's better to implement something like a new httpclient middleware `CanonicalizedHeadersMiddleware`, which will convert ALL headers to canonical format and inject it here after `SetHeadersMiddleware`: 

https://github.com/grafana/grafana/blob/9a41277ecdbf327fadb6acfbf0e3415098d6503c/pkg/services/pluginsintegration/clientmiddleware/oauthtoken_middleware.go#L95

& in other places if necessary.

If this is at the end of the middlewares chain, it will ensure that all outgoing http headers are canonical, regardless of what the caller has set.
If there are clashing names (like X-Id-Token and X-ID-Token both being present), we can make it give precedence to the that's already in canonical form.

Does this make sense to you? :grin: 

Correct me if I am wrong, but this middleware won't causes issues for plugin developers, because the httpclient middlewares are a completely different thing from the "headers" used by the plugins sdk, so backend plugins will still keep receiving `X-ID-Token` (so no breaking changes).

Reply:
Added what I have described above as WIP on this branch: https://github.com/grafana/grafana/tree/giuseppe/canonical-headers-middleware

If it makes sense I can close this PR, add tests and open a new PR for that branch, lmk what you think